### PR TITLE
PR: Change behavior of Variable Explorer editor buttons

### LINKED ISSUE #7156: Change behavior of the Variable Explorer editors buttons
The problem is right now the Ok button sends a serialized copy of the viewed object to the kernel, even if the user has not made changes to it.

### PR DESCRIPTION
**Pull request by dalthviz**
### Pull Request Checklist

* [x] Read and followed this repo's [Contributing Guidelines](https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md)
* [x] Based your PR on the latest version of the correct branch (master or 3.x)
* [x] Followed [PEP8](https://www.python.org/dev/peps/pep-0008/) for code style
* [x] Ensured your pull request hasn't eliminated unrelated blank lines/spaces,
      modified the ``spyder/defaults`` directory, or added new icons/assets
* [x] Wrote at least one-line docstrings for any new functions
* [x] Added at least one unit test covering the changes, if at all possible
* [x] Described your changes and the motivation for them below
* [x] Noted what issue(s) this pull request resolves, creating one if needed
* [x] Included a screenshot, if this PR makes any visible changes to the UI

## Description of Changes

Change behavior of default `OK` and `Cancel` buttons. Now `OK` is `Save and close` (disabled by default) and `Cancel` is the `Close` button.

### Issue(s) Resolved

Fixes #7156

### CODE REVIEW

**Comment by dalthviz** [comment]
@ccordoba12 a preview:

imagen

**Comment by jitseniesen** [comment]
I am not so sure about the names of the buttons. I believe that usually if you have an `OK` and an `Apply` button, the `Apply` button applies the changes without closing the window and the `OK` button applies the changes and closes the window, while here (if I understand it correctly), `OK` closes the window without applying the changes.

Maybe `Apply` and `Close` is better?

**Comment by ccordoba12** [comment]
> Maybe Apply and Close is better?

Yes, I think you're right. @dalthviz, please apply @jitseniesen's suggestion.

**Comment by CAM-Gerlach** [comment]
So, just to be clear, if I ``Apply`` the changes and then click ``Close``, the changes stay applied, correct? Also, the ``Close`` button has the exact same effect as the normal window close button (``X``), correct?

**Comment by CAM-Gerlach** [comment]
As @jitseniesen also stated, the universally followed UI standard is that ``OK`` saves/commits the changes and closes the dialog, while ``Apply`` saves/commits but doesn't close the dialog. Therefore, if what you've stated is the behavior this PR currently implements, either one of the two possibilities should presumably be applied—changing the name of the buttons, or their exact function: 

* The ``Apply`` button should read ``OK`` and close the dialog, to conform with the universal convention, and, really, the ``Close`` button changed back to ``Cancel`` since it is both clearer about its own function (both closing the dialog and not saving the changes) and how it is different from ``Close``, as well as following the standard idiomatic pattern users expect (OK/Cancel, Apply/Close)
* Or, what I would think is better and more in the spirit of this PR, keeping the names but making the ``Apply`` button save the changes but *not* close the dialog, while the ``Close`` button closes the dialog without applying any unapplied changes, but keeping anything that has been applied (as it does currently). This makes more sense, I think, with what this PR is trying to do, and is clearer to users what each does vs. the previous buttons.

**Comment by dalthviz** [comment]
@ccordoba12 `Apply` closes the dialog immediatly.

About the comments of @CAM-Gerlach maybe `OK` and `Cancel` are good or we could use something like `Save Changes` and `Cancel` to be explicit about the action of each button? The second suggestion is also valid, although it would need probably to override the closeEvent of the dialogs and some signal to trigger the save of the changes. What do you guys think @ccordoba12 @CAM-Gerlach @jitseniesen ?

**Comment by jitseniesen** [comment]
I prefer CAM-Gerlach's second suggestion.

As an alternative which is probably less work and fixes the problem with that applying the changes can be expensive, and it is an expensive no-op if the user did not make any changes (which is quite common). Is that correct ... how about we leave everything as it is, except that if the user does not make any changes and clicks `OK`, the editor does not send a serialized copy of the unchanged object to the kernel?

**Comment by CAM-Gerlach** [comment]
@dalthviz One option is keeping the current ``Apply`` button's function, renaming it ``OK``, and adding a another button named ``Apply`` that would save without closing, (like we have in our preferences and is pretty standard), but that last part might be outside the scope of this PR. 

> ``Save Changes`` 

Hmmm...well, in a way ``OK`` is actually clearer since it also makes clear that it will close the dialog; generally it is assumed that ``OK`` will save changes (as that is why it is distinct from ``Cancel``) and it is used everywhere for that purpose; e.g. in our very own preferences. Furthermore, ``OK`` and ``Cancel`` have better parallelism and clearer separation than ``Save Changes`` and ``Close``; while ``Apply`` and ``Close`` are also fairly clear.

> how about we leave everything as it is, except that if the user does not make any changes and clicks OK, the editor does not send a serialized copy of the unchanged object to the kernel?

Yeah; that's essentially what I was thinking of for my first option—simpler, though a little less clear/functional.

**Comment by CAM-Gerlach** [comment]
> I don't like Cancel because the real operation is simply closing the dialog. So the Close button doesn't need to be changed.

Right, but the problem with the ``Close`` wording is that its closing without saving changes; users could assume their changes are automatically saved. More essentially, ``Close`` and ``OK`` are just rather confusing together.

To make it clearer with minimum effort and keeping the ``Close`` wording on that button, how about we just rename the ``Apply`` button ``Save and Close`` instead? Then it makes it very clear and explicit what it does, both by itself and relative to the other button.

**Comment by ccordoba12** [comment]
> how about we just rename the Apply button Save and Close instead?

Ok, that's fine. @dalthviz, please implement this suggestion and also change the name of the variables that hold the editor buttons (e.g. `btn_apply` needs to be `btn_save_and_close`).

File: spyder/widgets/variableexplorer/arrayeditor.py [hunk 1]
```diff
@@ -605,6 +606,8 @@ def __init__(self, parent=None):
         self.arraywidget = None
         self.stack = None
         self.layout = None
+        self.btn_save_and_close = None
+        self.btn_ok = None
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/arrayeditor.py]: @dalthviz, this needs to be renamed to `btn_close`.

File: spyder/widgets/variableexplorer/collectionseditor.py [hunk 2]
```diff
@@ -1321,6 +1325,8 @@ def __init__(self, parent=None):
 
         self.data_copy = None
         self.widget = None
+        self.btn_save_and_close = None
+        self.btn_ok = None
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/collectionseditor.py]: Rename to `btn_close`

File: spyder/widgets/variableexplorer/collectionseditor.py [hunk 3]
```diff
@@ -1374,6 +1389,13 @@ def setup(self, data, title='', readonly=False, width=650, remote=False,
         # Make the dialog act as a window
         self.setWindowFlags(Qt.Window)
 
+    def save_and_close_enable(self):
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/collectionseditor.py]: Please add `@Slot()` above this line.

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 4]
```diff
@@ -722,15 +724,28 @@ def setup_and_check(self, data, title=''):
...
-        bbox.accepted.connect(self.accept)
-        bbox.rejected.connect(self.reject)
-        btn_layout.addWidget(bbox)
+
+        self.btn_save_and_close = QPushButton(_('Save and Close'))
+        self.btn_save_and_close.setDisabled(True)
+        self.btn_save_and_close.clicked.connect(self.accept)
+        btn_layout.addWidget(self.btn_save_and_close)
+
+        self.btn_ok = QPushButton(_('Close'))
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: Rename to `btn_close`

File: spyder/widgets/variableexplorer/dataframeeditor.py [hunk 5]
```diff
@@ -722,15 +724,28 @@ def setup_and_check(self, data, title=''):
...
+        self.btn_ok.setAutoDefault(True)
+        self.btn_ok.setDefault(True)
+        self.btn_ok.clicked.connect(self.reject)
+        btn_layout.addWidget(self.btn_ok)
 
         self.layout.addLayout(btn_layout, 2, 0)
 
         return True
 
+    def save_and_close_enable(self):
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/dataframeeditor.py]: This function needs to receive two parameters: `top_left` and `bottom_right` that account for the indexes passed by `dataChanged`.

Comment (ccordoba12) [reply spyder/widgets/variableexplorer/dataframeeditor.py]: Also, above this method you need to add

`@Slot(QModelIndex, QModelIndex)`

File: spyder/widgets/variableexplorer/arrayeditor.py [hunk 6]
```diff
@@ -737,9 +752,17 @@ def setup_and_check(self, data, title='', readonly=False,
         self.setWindowFlags(Qt.Window)
         
         return True
-            
+
+    def save_and_close_enable(self):
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/arrayeditor.py]: Above this method you need to add

    @Slot(QModelIndex, QModelIndex)

Comment (ccordoba12) [reply spyder/widgets/variableexplorer/arrayeditor.py]: And this needs two parameters: `top_left` and `bottom_right`.

File: spyder/widgets/variableexplorer/texteditor.py [hunk 7]
```diff
@@ -50,23 +52,31 @@ def __init__(self, text, title='', font=None, parent=None,
...
-        bbox.accepted.connect(self.accept)
-        bbox.rejected.connect(self.reject)
-        self.layout.addWidget(bbox)
-        
+            self.btn_save_and_close = QPushButton(_('Save and Close'))
+            self.btn_save_and_close.setDisabled(True)
+            self.btn_save_and_close.clicked.connect(self.accept)
+            btn_layout.addWidget(self.btn_save_and_close)
+
+        self.btn_ok = QPushButton(_('Close'))
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/texteditor.py]: Rename to `btn_close`

File: spyder/widgets/variableexplorer/texteditor.py [hunk 8]
```diff
@@ -50,23 +52,31 @@ def __init__(self, text, title='', font=None, parent=None,
 
         # Text edit
         self.edit = QTextEdit(parent)
-        self.edit.textChanged.connect(self.text_changed)
         self.edit.setReadOnly(readonly)
+        self.edit.textChanged.connect(self.text_changed)
```
Comment (ccordoba12) [inline spyder/widgets/variableexplorer/texteditor.py]: Please add `@Slot()` above the definition of the `text_changed` method.

**Pull request merged**